### DISCUSSION TITLE
Is io_uring/liburing completely compatible with kTLS?

### DISCUSSION THREAD

**Discussion by cmazakas** [axboe/liburing#968] [Q&A]
I'm looking into kernel TLS and I'm wondering if it works with the io_uring stuff like direct descriptors and `io_uring_prep_connect()` and such.

I see from the API docs that one does a lot of this stuff via `setsockopt()`. Is there a way to make this async with io_uring so I can do a SQE chain of:
```
io_uring_prep_socket_direct() -> setsocketopt() -> io_uring_prep_connect()
```
?

**Comment by davidzeng0** [comment]
No, because io_uring has no setsockopt opcode. You can however `prep_socket` -> `setsockopt` -> `files_update`. You must create a non-direct file because setsockopt takes file descriptors, not io_uring direct file indexes. Once setup is done, you can send it to io_uring and then do all your work there

**Comment by axboe** [comment]
io_uring will have set/getsockopt shortly, it's staged for the 6.7 kernel:

https://git.kernel.dk/cgit/linux/log/?h=for-6.7/io_uring-sockopt

until that is in, the solution is indeed to do all of that first and then register the file.
